### ISSUE TITLE
multiplayer issue

### ISSUE THREAD

**Issue by gabriel-tandil** [Anuken/Mindustry#10329]
### Platforms

Android

### Build

release build 146

### Issue

I found it reported as #10106 but it was closed due to insufficient data, I provide more information and if you need more data or to do some testing I am available and well qualified to do it.

It happened to me on both android and linux platforms.

Other players see my server, and although the screen initially shows that there is an error connecting, they are able to connect. In my game I see them immobile but they see the map but do not see their player. 

Another piece of information... Playing the same game on the local network works, but if even from the machine that has the server I try to connect as a game on the internet then it fails.

I open my game for multiplayer

I have the port redirected in my router correctly.

https://github.com/user-attachments/assets/9dbbc7d0-7415-4f10-bc88-a5829faf6ec6

### Steps to reproduce

1. start server (can be the standalone server or from the game)
2. from another machine try to connect to the game (can be by ip or by domain)
3. the map loads but the player is not visible.

### Mods used

none

### Save file

here is
[multiplayerError.zip](https://github.com/user-attachments/files/17775924/multiplayerError.zip)

### (Crash) logs

none

### Submission

- [X] I have updated to the latest release (https://github.com/Anuken/Mindustry/releases) to make sure my issue has not been fixed.
- [X] I have searched the closed and open issues to make sure that this problem has not already been reported.
- [X] I am not using Foo's Client, and have made sure the bug is not caused by mods I have installed.

**Comment by camelStyleUser** [comment]
what port forwarding rule do you have

**Comment by camelStyleUser** [comment]
is it UDP+TCP 6567->6567(or some other mindustry server ports)

**Comment by gabriel-tandil** [comment]
> is it UDP+TCP 6567->6567(or some other mindustry server ports)
@camelStyleUser exactly
![imagen](https://github.com/user-attachments/assets/d4b136f5-dcfd-466d-ba5b-65a6cf8a9100)

**Comment by Anuken** [comment]
Does this issue still occur in the [bleeding edge builds?](https://github.com/Anuken/MindustryBuilds/releases)

**Comment by Anuken** [comment]
Additionally, are you capable of building the game yourself for testing, or would you rather I build them myself/commit to master for fetching BE builds?

**Comment by gabriel-tandil** [comment]
> Does this issue still occur in the [bleeding edge builds?](https://github.com/Anuken/MindustryBuilds/releases)

Tnx! Next week I'll try that


> Additionally, are you capable of building the game yourself for testing, or would you rather I build them myself/commit to master for a BE build?

i guess I can build th game by myself, I'll try

**Comment by SITUVNgcd** [comment]
This is look like `Vars.logic.reset()` on server.

**Comment by gabriel-tandil** [comment]
@Anuken I already tested with the latest version listed in _bleeding edge builds_ and the problem occurs the same way.

I still couldn't successfully build the game from the **_Anuken/Mindustry_** git but I'm working on it.... Any idea what I can verify with a version built from there when I succeed?

**Comment by gabriel-tandil** [comment]
I can now make my custom build, I remain attentive.

**Comment by gabriel-tandil** [comment]
here is
![imagen](https://github.com/user-attachments/assets/fd88856d-8690-49e7-8caf-4affbc3ad86f)

**Comment by gabriel-tandil** [comment]
here is
![imagen](https://github.com/user-attachments/assets/fd88856d-8690-49e7-8caf-4affbc3ad86f)

**Comment by way-zer** [comment]
UDP stream from server to client  is fail.

**Comment by gabriel-tandil** [comment]
> UDP stream from server to client is fail.

@way-zer I don't think it's that...
try the udp communication through the port with 
`$ nc -ul -p 6567`
on the machine where the server would run, and 
`$ echo “test udp” | nc -u gabrieltandil.ddns.net 6567`
on another machine with another internet connection and the data goes fine.

**Comment by way-zer** [comment]
> The strange thing is that the connection _requires_ functioning UDP to work properly; otherwise, it would fail to connect with a `timed out during UDP registration` message. Something else is going on here.

The handshake process only involves `TCP C->S` `TCP S->C` `UDP C->S`, no `UDP S->C`.

**Comment by way-zer** [comment]
> > UDP stream from server to client is fail.
> 
> @way-zer I don't think it's that... I try the UDP communication through the port with `$ nc -ul -p 6567` on the machine where the server would run, and `$ echo “test udp” | nc -u gabrieltandil.ddns.net 6567` on a machine with a different internet connection and the data flows correctly.

This is still test oneway, from `C->S`, not `S->C`

**Comment by gabriel-tandil** [comment]
Thanks for your answer @way-zer 

Then surely the problem comes from that side if bidirectional UTP flow is expected.

A possible first improvement would be to inform the user about the specific error with some message.

Anyway I think you should consider whether using UTP “bare metal” is a good idea for bidirectional communication, since not being connection oriented these problems can occur by NATs in the middle for example.

One possibility would be to request that the clients also have an open UDP port on their machine and router, although it would make the configuration more complicated for the players.

**Comment by way-zer** [comment]
Under normal circumstances, NAT will maintain port mappings, even for UDP. Since many programs only use UDP and require two-way communication, this is not a problem. So, I think this is not a problem of Mindustry.

For Mindustry, adding UDP verification in the handshake phase will lead to incompatibility. In addition, UDP is inherently unreliable, which may increase the possibility of handshake failure and cause confusion.

> the server list it shows that there is an error
Server list uses only UDP for ping, There is a problem with your UDP connection, so an error will be displayed.

**Comment by way-zer** [comment]
This is an individual problem, you may need a dedicated server, even I have encountered it, because of the operator's poor QoS policy. 

I think this issue can be closed.

**Comment by gabriel-tandil** [comment]
Well, if you want to close it that's fine.
I think it's a problem that affects several people and can be improved (I actually saw this reported in other tickets in the past). Many ISPs have similar policies and although not ideal it is possible to use them.

Anyway I opened the ticket with the hope that my comments and tests will serve to improve the game, which is very good anyway. I don't know if I can contribute anything else.

**Comment by way-zer** [comment]
It's not easy to improve. Many games use UDP and the connection can't be pure TCP, otherwise the congestion problem will be severe.

Just adding one error does not improve the user experience.

**Comment by gabriel-tandil** [comment]
I understand, that's why I suggested to open a port in the client's router as well, although I understand that this solution may not be good for everyone.

But I leave you the concern, it is a subject that can be improved although it may not be simple, while vpn service vendors at least have their business :laughing: 

Anyway I think that a more detailed error would be helpful for users. 🙏

**Comment by way-zer** [comment]
> open a port in the client's router

client's router has no reason to drop the packet, opening a port don't mean you can receive packets. So client can't do anything for this, also the game itself.

**Comment by gabriel-tandil** [comment]
I understand your point, it is just a possible solution for something that can often fail. Anyway thank you for your attention and your answers.

**Comment by gabriel-tandil** [comment]
> What about the confusion caused by situations such as this, where the UDP S->C connection doesn't actually work? Isn't it better to fail early if the server can't establish a proper connection?

Yes, I think that showing a fault (and if possible adding some details of the cause) would be the best thing to do.

**Comment by way-zer** [comment]
> What about the confusion caused by situations such as this, where the UDP S->C connection doesn't actually work? Isn't it better to fail early if the server can't establish a proper connection?

You can add verification, but it's better to keep resending UDP periodically (1s) before the timeout.

**Comment by lyl-1520202** [comment]
Kiedy powstaną guzy i gwiazdy E?

**Comment by lyl-1520202** [comment]
Kiedy powstaną guzy i gwiazdy E?

**Comment by Mavment** [comment]
Btw this bug also acures with global servers sometimes, like it happaned 2-3 with me on pc and this happanes very often on android, it depends on internet, for example when i had slow internet this acured very often (~25% i tried to join), with good internet connection this happanes very rare, probably useless comment but anyways
